[AMDGPU][MC] GFX9 - allow op_sel in v_interp_p2_f16

jwanggit86 opened the pull request:
AMDGPU documentation states op_sel[3] can be used in v_interp_p2_f16 in GFX9.

jwanggit86 committed:
[AMDGPU][MC] GFX9 - allow op_sel in v_interp_p2_f16

AMDGPU documentation states op_sel[3] can be used in
v_interp_p2_f16 in GFX9.

Sisyph commented on llvm/test/MC/AMDGPU/vop3-gfx9.s:
```diff
@@ -566,6 +566,36 @@ v_interp_p2_f16 v5, v2, attr0.x, v3 clamp
 // NOSICI: :[[@LINE-2]]:{{[0-9]+}}: error: instruction not supported on this GPU
 // VI: v_interp_p2_f16 v5, v2, attr0.x, v3 clamp ; encoding: [0x05,0x80,0x76,0xd2,0x00,0x04,0x0e,0x04]
 
+v_interp_p2_f16 v5, v2, attr0.x, v3 op_sel:[1]
```
I think the convention here is to pass whatever op_sel bits were input to the output, not drop them. For comparison, sp3 assembler takes whatever opsel bits you specified, and puts them in the output. We do the same for v_pack_b32_f16 on gfx9. On newer targets, especially in true16 mode, we tend to give an assembler error when opsel bit is not supported on a certain. But what this does, just dropping the bits that are non-functional, is not what we usually do.

jwanggit86 committed:
This commit:

(1) In parsing, encode op_sel bits in src0_modifiers and src2_modifiers.
(2) In printing, let printPackedModier() handle it, but had to modify
that function a little.

jwanggit86 commented on llvm/lib/Target/AMDGPU/MCTargetDesc/AMDGPUInstPrinter.cpp:
```diff
@@ -1299,6 +1299,14 @@ void AMDGPUInstPrinter::printOpSel(const MCInst *MI, unsigned,
       O << " op_sel:[" << FI << ',' << BC << ']';
     return;
   }
+  if (Opc == AMDGPU::V_INTERP_P2_F16_opsel_gfx9) {
```
Still, because there's no src1, the printed op_sel[] only contains 3 elements, even when the input has 4 elements.

jwanggit86 committed:
Fix code format

Sisyph commented on llvm/lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp:
```diff
@@ -9310,6 +9310,38 @@ void AMDGPUAsmParser::cvtVOP3Interp(MCInst &Inst, const OperandVector &Operands)
+    int OpSelIdx = AMDGPU::getNamedOperandIdx(Opc, AMDGPU::OpName::op_sel);
+    unsigned OpSel = Inst.getOperand(OpSelIdx).getImm();
+
+    const AMDGPU::OpName Ops[] = {AMDGPU::OpName::src0, AMDGPU::OpName::src1,
```
This looks copied from cvtVINTERP. Can you outline it into a helper method?

Sisyph commented on llvm/test/MC/AMDGPU/vop3-gfx9.s:
```diff
@@ -566,6 +566,86 @@ v_interp_p2_f16 v5, v2, attr0.x, v3 clamp
+// NOSICI: :[[@LINE-2]]:{{[0-9]+}}: error: instruction not supported on this GPU
+// NOVI: :[[@LINE-3]]:{{[0-9]+}}: error: not a valid operand.
+
+v_interp_p2_f16 v5, v2, attr0.x, v3 op_sel:[1,0,1,1]
```
The fact that this does not round trip from assembly to disassembly will be pretty confusing for users I think. @rampitec What do you think?

jwanggit86 replied:
As far as I know, round-trip already is not always guaranteed for op_sel, e.g. when the input has extraneous bits. Here is an example:
```
$ cat z.s
v_pk_max_f16 v5, v1, v2 op_sel:[1,1,1]
v_pk_max_f16 v5, v1, v2 op_sel:[1,1,1,1]

$ ../../build/bin/llvm-mc -triple=amdgcn -mcpu=gfx908 -show-encoding z.s
        v_pk_max_f16 v5, v1, v2 op_sel:[1,1]    ; encoding: [0x05,0x58,0x92,0xd3,0x01,0x05,0x02,0x18]
        v_pk_max_f16 v5, v1, v2 op_sel:[1,1]    ; encoding: [0x05,0x58,0x92,0xd3,0x01,0x05,0x02,0x18]

$ cat z.txt
0x05,0x58,0x92,0xd3,0x01,0x05,0x02,0x18

$ ../../build/bin/llvm-mc -triple=amdgcn -mcpu=gfx908 -show-encoding -disassemble z.txt
        v_pk_max_f16 v5, v1, v2 op_sel:[1,1]    ; encoding: [0x05,0x58,0x92,0xd3,0x01,0x05,0x02,0x18]
```
Here only two bits are expected in op_sel, but the input instructions have more. The extra bits cannot be encoded. So the assembler only encodes the two valid bits, and the printer only prints two bits. The disassembler only sees two bits, and only prints two bits.

One option is for the assembler to reject the instruction when there are too many op_sel bits, which is a non-trivial change. Do you think we should do this? @rampitec @Sisyph

arsenm replied:
We don't round trip lots of extraneous constructs, this isn't unique. I don't think it's worth putting extra effort into preserving defaulted junk bits. If they are set to non-default values, I'd almost expect them to pass through

jwanggit86 replied:
In the given example, the input and output are different because the input is incorrect with 4 bits while the instruction only takes 3. If we don't want to do the extra work of checking the number of op_sel bits, then the extraneous bit is silently ignored, which is what happened in the given example, and, as pointed out above, is also what happens elsewhere with op_sel.

arsenm commented on llvm/lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp:
```diff
@@ -9310,6 +9310,38 @@ void AMDGPUAsmParser::cvtVOP3Interp(MCInst &Inst, const OperandVector &Operands)
     addOptionalImmOperand(Inst, Operands, OptionalIdx,
                           AMDGPUOperand::ImmTyOModSI);
+
+  // Some v_interp instrutions use op_sel[3] for dst.
```
```suggestion
  // Some v_interp instructions use op_sel[3] for dst.
```

jwanggit86 committed:
Create a helper method for common code in two functions

shiltian commented on llvm/lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp:
```diff
@@ -9239,6 +9178,32 @@ static bool isRegOrImmWithInputMods(const MCInstrDesc &Desc, unsigned OpNum) {
 
+void AMDGPUAsmParser::cvtOpSelHelper(MCInst &Inst, unsigned OpSel) {
+  unsigned Opc = Inst.getOpcode();
+  const AMDGPU::OpName Ops[] = {AMDGPU::OpName::src0, AMDGPU::OpName::src1,
+                                AMDGPU::OpName::src2};
+  const AMDGPU::OpName ModOps[] = {AMDGPU::OpName::src0_modifiers,
+                                   AMDGPU::OpName::src1_modifiers,
+                                   AMDGPU::OpName::src2_modifiers};
```
nit: I'd just merge these two, and then use `i` and `i + 3` if needed.

Also, `constexpr` can be used.

shiltian commented on llvm/lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp:
```diff
@@ -1944,6 +1928,7 @@ class AMDGPUAsmParser : public MCTargetAsmParser {
 
   void cvtVOP3Interp(MCInst &Inst, const OperandVector &Operands);
   void cvtVINTERP(MCInst &Inst, const OperandVector &Operands);
+  void cvtOpSelHelper(MCInst &Inst, unsigned OpSel);
```
It is not clear to me what this helper does. Can you make the function name more clear, and add comment?

jwanggit86 replied:
It converts the raw op_sel value in the asm instruction to an MCInst operand. This function was created to avoid code duplication in two functions (the two call sites).

Sisyph reviewed (commented):
Can you please add a few assembly tests with 3 values in op_sel? like `'... op_sel:[1,0,1]'`

With that LGTM

jwanggit86 committed:
Printer should print 4 bits for op_sel

jwanggit86 committed:
Fix code formatting

jwanggit86 commented:
In the printer, check for this special case, i.e., instructions with src0, src2 but no src1, and ensure 4 bits for op_sel are printed. For src1 print the default value.

jwanggit86 commented:
> Can you please add a few assembly tests with 3 values in op_sel? like `'... op_sel:[1,0,1]'`
> 
> With that LGTM

Added 3 bits tests.

Sisyph reviewed (approved):
Thanks, LGTM
